lua/complementree/utils.lua

```diff
@@ -9,4 +13,73 @@ function M.cword(complete_item)
   return (complete_item.word or complete_item.abbr)
 end
 
+M.make_relative_path = function(path, root)
+  local baselen = #root
+  if path:sub(0, baselen) == root then
+    path = path:sub(baselen + 2)
+  end
+  return path
+end
+
+local scandir_default_opts = {
+  ignore_hidden = true,
+  max_depth = 8,
+  match_patterns = {},
+}
+
+M.scan_dir = function(root_dir, opts, list, depth)
+  opts = opts or scandir_default_opts
+  list = list or {}
+  depth = depth or 1
+
+  local function is_ignored_path(path)
+    return (opts.ignore_hidden and path:sub(1, 1) == '.')
+  end
+
+  local function is_matched_filetype(path)
+    for _, pattern in ipairs(opts.match_patterns) do
+      if path:match(pattern) then
+        return true
+      end
+    end
+    return false
+  end
+
+  local req = uv.fs_scandir(root_dir)
+  if not req then
+    return list
+  end
+
+  local function iter()
+    return uv.fs_scandir_next(req)
+  end
+
+  local sub_dirs = {}
+  for path, ftype in iter do
+    if ftype == 'directory' then
+      if not is_ignored_path(path) then
+        sub_dirs[#sub_dirs + 1] = path
+      end
+    elseif ftype == 'file' then
+      if not is_ignored_path(path) then
+        if opts.match_patterns ~= {} and not is_matched_filetype(path) then
+          goto continue
+        end
+        list[#list + 1] = ('%s%s%s'):format(root_dir, os_sep, path)
+        ::continue::
```

Comment:
Why not inverting it ?
```suggestion
        if opts.match_patterns == {} or is_matched_filetype(path) then
                list[#list + 1] = ('%s%s%s'):format(root_dir, os_sep, path)
        end
```

It avoids some complexities.